src/type/matrix/DenseMatrix.js

```diff
@@ -524,6 +525,71 @@ export const createDenseMatrixClass = /* #__PURE__ */ factory(name, dependencies
     return this._size.slice(0) // return a clone of _size
   }
 
+  /**
+   * Applies a callback function to a reference to each element of the matrix
+   * @memberof DenseMatrix
+   * @param {Function} callback   The callback function is invoked with three
+   *                              parameters: an array, an integer index to that
+   *                              array, and the Matrix being traversed.
+   *
+   * @return {DenseMatrix} matrix
+   */
+  DenseMatrix.prototype._forEach = function (callback) {
+    // matrix instance
+    const me = this
+    const s = me.size()
+
+    // if there is only one dimension, just loop through it
+    if (s.length === 1) {
+      for (let i = 0; i < s[0]; i++) {
+        callback(me._data, i, [i])
+      }
+      return
+    }
+
+    // keep track of the current index permutation
+    const index = Array(s.length).fill(0)
+
+    // store a reference of each dimension of the matrix for faster access
+    const data = Array(s.length - 1)
+    const last = data.length - 1
+
+    data[0] = me._data[0]
+    for (let i = 0; i < last; i++) {
+      data[i + 1] = data[i][0]
+    }
+
+    index[last] = -1
+    while (true) {
+      let i
+      for (i = last; i >= 0; i--) {
+        // march index to the next permutation
+        index[i]++
+        if (index[i] === s[i]) {
+          index[i] = 0
+          continue
+        }
+
+        // update references to matrix dimensions
+        data[i] = i === 0 ? me._data[index[i]] : data[i - 1][index[i]]
+        for (let j = i; j < last; j++) {
+          data[j + 1] = data[j][0]
+        }
+
+        // loop through the last dimension and map each value
+        for (let j = 0; j < s[data.length]; j++) {
+          index[data.length] = j
+          callback(data[last], j, index)
```

Comment:
Can you please change this to return a copy of index like `callback(data[last], j, index.slice())`, and remove the `index.slice(0)` from the `forEach` method?

There are currently no unit tests in place that test whether the index is immutable. Can you please adjust the following two unit tests?

1. [DenseMatrix > map > should invoke callback with parameters value, index, obj](https://github.com/josdejong/mathjs/blob/2a4faaa03cebaa77b9012145c03b07db0673bd94/test/unit-tests/type/matrix/DenseMatrix.test.js#L733-L756)

    Remove `JSON.stringify(...)` around the created array in the callback and change the expected output to be arrays instead of strings.

2. [DenseMatrix > forEach > should invoke callback with parameters value, index, obj](https://github.com/josdejong/mathjs/blob/2a4faaa03cebaa77b9012145c03b07db0673bd94/test/unit-tests/type/matrix/DenseMatrix.test.js#L790-L806)

    Remove `math.clone(...)` around the created array in the callback.